pallets/liquidity-pools/src/tests.rs

```diff
@@ -0,0 +1,622 @@
+	permissions::{PermissionScope, PoolRole, Role},
+	tokens::{AssetMetadata, CrossChainTransferability, CurrencyId, CustomMetadata},
+};
+use cfg_utils::vec_to_fixed_array;
+use frame_support::{
+	assert_noop, assert_ok,
+	traits::{fungibles::Mutate as _, PalletInfo as _},
+};
+use sp_runtime::{DispatchError, TokenError};
+use staging_xcm::{
+	v4::{Junction::*, Location, NetworkId},
+	VersionedLocation,
+};
+
+use crate::{mock::*, Error, GeneralCurrencyIndexOf, Message};
+
+const CHAIN_ID: u64 = 1;
+const ALICE: AccountId = AccountId::new([0; 32]);
+const CONTRACT_ACCOUNT: [u8; 20] = [1; 20];
+const CONTRACT_ACCOUNT_ID: AccountId = AccountId::new([1; 32]);
+const EVM_ADDRESS: DomainAddress = DomainAddress::EVM(CHAIN_ID, CONTRACT_ACCOUNT);
+const AMOUNT: Balance = 100;
+const CURRENCY_ID: CurrencyId = CurrencyId::ForeignAsset(1);
+const POOL_ID: PoolId = 1;
+const TRANCHE_ID: TrancheId = [1; 16];
+const NOW: Millis = 0;
+const NAME: &[u8] = b"Token name";
+const SYMBOL: &[u8] = b"Token symbol";
+const DECIMALS: u8 = 6;
+
+mod util {
+	use super::*;
+
+	pub fn default_metadata() -> AssetMetadata {
+		AssetMetadata {
+			decimals: DECIMALS as u32,
+			name: Vec::from(NAME).try_into().unwrap(),
+			symbol: Vec::from(SYMBOL).try_into().unwrap(),
+			..cfg_types::tokens::default_metadata()
+		}
+	}
+
+	pub fn transferable_metadata() -> AssetMetadata {
+		AssetMetadata {
+			additional: CustomMetadata {
+				transferability: CrossChainTransferability::LiquidityPools,
+				..Default::default()
+			},
+			..default_metadata()
+		}
+	}
+
+	pub fn wrapped_transferable_metadata() -> AssetMetadata {
+		let pallet_index = PalletInfo::index::<LiquidityPools>();
+		AssetMetadata {
+			location: Some(VersionedLocation::V4(Location::new(
+				0,
+				[
+					PalletInstance(pallet_index.unwrap() as u8),
+					GlobalConsensus(NetworkId::Ethereum { chain_id: CHAIN_ID }),
+					AccountKey20 {
+						network: None,
+						key: CONTRACT_ACCOUNT,
+					},
+				],
+			))),
+			..transferable_metadata()
+		}
+	}
+
+	pub fn currency_index(currency_id: CurrencyId) -> u128 {
+		GeneralCurrencyIndexOf::<Runtime>::try_from(currency_id)
+			.unwrap()
+			.index
+	}
+}
+
+mod transfer {
+	use super::*;
+
+	#[test]
+	fn success() {
+		System::externalities().execute_with(|| {
+			AssetRegistry::mock_metadata(|_| Some(util::wrapped_transferable_metadata()));
+			TransferFilter::mock_check(|_| Ok(()));
+			Tokens::mint_into(CURRENCY_ID, &ALICE, AMOUNT).unwrap();
+			Gateway::mock_submit(|sender, destination, msg| {
+				assert_eq!(sender, ALICE);
+				assert_eq!(destination, EVM_ADDRESS.domain());
+				assert_eq!(
+					msg,
+					Message::Transfer {
+						currency: util::currency_index(CURRENCY_ID),
+						sender: ALICE.into(),
+						receiver: EVM_ADDRESS.address(),
+						amount: AMOUNT
+					}
+				);
+				Ok(())
+			});
+
+			assert_ok!(LiquidityPools::transfer(
+				RuntimeOrigin::signed(ALICE),
+				CurrencyId::ForeignAsset(1),
+				EVM_ADDRESS,
+				AMOUNT
+			));
```

Comment:
Initially I removed that check because it's checking that orml_tokens is correctly working (which is out of scope from this UT test), but thinking again, it makes sense to check that actually this method calls correctly to orml_tokens, with the correct amount. So, will do 👍🏻

Reply:
Very good point BTW